man/man1/notmuch-tag.1
.TH NOTMUCH-TAG 1 2013-12-30 "Notmuch 0.17"
.SH NAME
notmuch-tag \- add/remove tags for all messages matching the search terms

.SH SYNOPSIS
.B notmuch tag
.RI [ options "...] +<" tag ">|\-<" tag "> [...] [\-\-] <" search-term "> [...]"

.B notmuch tag
.RI "--batch"
.RI "[ --input=<" filename "> ]"


.SH DESCRIPTION

Add/remove tags for all messages matching the search terms.

See \fBnotmuch-search-terms\fR(7)
for details of the supported syntax for
.RI < search-term >.

Tags prefixed by '+' are added while those prefixed by '\-' are
removed. For each message, tag changes are applied in the order they
appear on the command line.

The beginning of the search terms is recognized by the first
argument that begins with neither '+' nor '\-'. Support for
an initial search term beginning with '+' or '\-' is provided
by allowing the user to specify a "\-\-" argument to separate
the tags from the search terms.

.B "notmuch tag"
updates the maildir flags according to tag changes if the
.B "maildir.synchronize_flags"
configuration option is enabled. See \fBnotmuch-config\fR(1) for
details.

Supported options for
.B tag
include
.RS 4
.TP 4
.BR \-\-remove\-all

Remove all tags from each message matching the search terms before
applying the tag changes appearing on the command line. This means
setting the tags of each message to the tags to be added. If there are
no tags to be added, the messages will have no tags.
.RE

.RS 4
.TP 4
.BR \-\-batch

Read batch tagging operations from a file (stdin by default). This is more
efficient than repeated
.B notmuch tag
invocations. See
.B TAG FILE FORMAT
below for the input format. This option is not compatible with
specifying tagging on the command line.
.RE

.RS 4
.TP 4
.BR "\-\-input=" <filename>

Read input from given file, instead of from stdin. Implies
.BR --batch .

.SH TAG FILE FORMAT

The input must consist of lines of the format:

.RI "+<" tag ">|\-<" tag "> [...] [\-\-] <" query ">"

Each line is interpreted similarly to
.B notmuch tag
command line arguments. The delimiter is one or more spaces ' '. Any
characters in
.RI < tag >
.B may
be hex-encoded with %NN where NN is the hexadecimal value of the
character. To hex-encode a character with a multi-byte UTF-8 encoding,
hex-encode each byte.
Any spaces in <tag>
.B must
be hex-encoded as %20. Any characters that are not
part of
.RI  < tag >
.B must not
be hex-encoded.

In the future tag:"tag with spaces" style quoting may be supported for
.RI < tag >
as well;
for this reason all double quote characters in
.RI < tag >
.B should
be hex-encoded.

The
.RI < query >
should be quoted using Xapian boolean term quoting rules: if a term
contains whitespace or a close paren or starts with a double quote, it
must be enclosed in double quotes (not including any prefix) and
double quotes inside the term must be doubled (see below for
examples).

Leading and trailing space ' ' is ignored. Empty lines and lines
beginning with '#' are ignored.

.SS EXAMPLE

The following shows a valid input to batch tagging. Note that only the
isolated '*' acts as a wildcard. Also note the two different quotings
of the tag
.B space in tags
.
.RS
.nf
+winner *
+foo::bar%25 -- (One and Two) or (One and tag:winner)
+found::it -- tag:foo::bar%
# ignore this line and the next

+space%20in%20tags -- Two
# add tag '(tags)', among other stunts.
+crazy{ +(tags) +&are +#possible\e -- tag:"space in tags"
+match*crazy -- tag:crazy{
+some_tag -- id:"this is ""nauty)"""
.fi
.RE

.SH SEE ALSO

\fBnotmuch\fR(1), \fBnotmuch-config\fR(1), \fBnotmuch-count\fR(1),
\fBnotmuch-dump\fR(1), \fBnotmuch-hooks\fR(5),
\fBnotmuch-insert\fR(1), \fBnotmuch-new\fR(1),
\fBnotmuch-reply\fR(1), \fBnotmuch-restore\fR(1),
\fBnotmuch-search\fR(1), \fBnotmuch-search-terms\fR(7),
\fBnotmuch-show\fR(1),


NEWS
Notmuch 0.17 (2013-12-30)
=========================

Incompatible change in SHA1 computation
---------------------------------------

Previously on big endian architectures like sparc and powerpc the
computation of SHA1 hashes was incorrect. This meant that messages
with overlong or missing message-ids were given different computed
message-ids than on more common little endian architectures like i386
and amd64.  If you use notmuch on a big endian architecture, you are
strongly advised to make a backup of your tags using `notmuch dump`
before this upgrade. You can locate the affected files using something
like:

    notmuch dump | \
      awk '/^notmuch-sha1-[0-9a-f]{40} / \
        {system("notmuch search --exclude=false --output=files id:" $1)}'

Command-Line Interface
----------------------

New options to better support handling duplicate messages

  If more than one message file is associated with a message-id,
  `notmuch search --output=files` will print all of them. A new
  `--duplicate=N` option can be used to specify which duplicate to
  print for each message.

  `notmuch count` now supports `--output=files` option to output the
  number of files associated with matching messages. This may be
  bigger than the number of matching messages due to duplicates
  (i.e. multiple files having the same message-id).

Improved `notmuch new` performance for unchanged folders

  `notmuch new` now skips over unchanged folders more efficiently,
  which can substantially improve the performance of checking for new
  mail in some situations (like NFS-mounted Maildirs).

`notmuch reply --format=text` RFC 2047-encodes headers

  Previously, this used a mix of standard MIME encoding for the reply
  body and UTF-8 for the headers.  Now, the text format reply template
  RFC 2047-encodes the headers, making the output a valid RFC 2822
  message.  The JSON/sexp format is unchanged.

`notmuch compact` command

  The new `compact` command exposes Xapian's compaction
  functionality through a more convenient interface than
  `xapian-compact`. `notmuch compact` will compact the database to a
  temporary location, optionally backup the original database, and
  move the compacted database into place.

Emacs Interface
---------------

`notmuch-tree` (formerly `notmuch-pick`) has been added to mainline

  `notmuch-tree` is a threaded message view for the emacs
  interface. Each message is one line in the results and the thread
  structure is shown using UTF-8 box drawing characters (similar to
  Mutt's threaded view). It comes between search and show in terms of
  amount of output and can be useful for viewing both single threads
  and multiple threads.

  Using `notmuch-tree`

  The main key entries to notmuch tree are

  'z' enter a query to view using notmuch tree (works in hello,
      search, show and tree mode itself)

  'Z' view the current query in tree notmuch tree (works from search
      and show)

  Once in tree mode, keybindings are mostly in line with the rest of
  notmuch and are all viewable with '?' as usual.

  Customising `notmuch-tree`

  `notmuch-tree` has several customisation variables. The most
  significant is the first notmuch-tree-show-out which determines the
  behaviour when selecting a message (with RET) in tree view. By
  default tree view uses a split window showing the single message in
  the bottom pane. However, if this option is set then it views the
  whole thread in the complete window jumping to the selected message
  in the thread. In either case command-prefix selects the other option.

Tagging threads in search is now race-free

  Previously, adding or removing a tag from a thread in a search
  buffer would affect messages that had arrived after the search was
  performed, resulting in, for example, archiving messages that were
  never seen.  Tagging now affects only the messages that were in the
  thread when the search was performed.

`notmuch-hello` refreshes when switching to the buffer

  The hello buffer now refreshes whenever you switch to the buffer,
  regardless of how you get there.  You can disable automatic
  refreshing by customizing `notmuch-hello-auto-refresh`.

Specific mini-buffer prompts for tagging operations

  When entering tags to add or remove, the mini-buffer prompt now
  indicates what operation will be performed (e.g., "Tag thread", "Tag
  message", etc).

Built-in help improvements

  Documentation for many commands has been improved, as displayed by
  `notmuch-help` (usually bound to "?").  The bindings listed by
  `notmuch-help` also now include descriptions of prefixed commands.

Quote replies as they are displayed in show view

  We now render the parts for reply quoting the same way they are
  rendered for show. At this time, the notable change is that replies
  to text/calendar are now pretty instead of raw vcalendar.

Fixed inconsistent use of configured search order

  All ways of interactively invoking search now honor the value of
  `notmuch-search-oldest-first`.

Common keymap for notmuch-wide bindings

  Several key bindings have been moved from mode-specific keymaps to
  the single `notmuch-common-keymap`, which is inherited by each
  notmuch mode.  If you've customized your key bindings, you may want
  to move some of them to the common keymap.

The `notmuch-tag` function now requires a list of tag changes

  For users who have scripted the Emacs interface: the `notmuch-tag`
  API has changed.  Previously, it accepted either a list of tag
  changes or a space-separated string of tag changes.  The latter is
  no longer supported and the function now returns nothing.

Fixed `notmuch-reply` putting reply in primary selection

  On emacs 24 notmuch-reply used to put the cited text into the
  primary selection (which could lead to inadvertently pasting this
  cited text elsewhere). Now the primary-selection is not changed.

Fixed `notmuch-show` invisible part handling


emacs/notmuch-tag.el
;; notmuch-tag.el --- tag messages within emacs
;;
;; Copyright © Damien Cassou
;; Copyright © Carl Worth
;;
;; This file is part of Notmuch.
;;
;; Notmuch is free software: you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; Notmuch is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;; General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with Notmuch.  If not, see <http://www.gnu.org/licenses/>.
;;
;; Authors: Carl Worth <cworth@cworth.org>
;;          Damien Cassou <damien.cassou@gmail.com>
;;
;;; Code:
;;

(require 'cl)
(require 'crm)
(require 'notmuch-lib)

(defcustom notmuch-tag-formats
  '(("unread" (propertize tag 'face '(:foreground "red")))
    ("flagged" (propertize tag 'face '(:foreground "blue"))
     (notmuch-tag-format-image-data tag (notmuch-tag-star-icon))))
  "Custom formats for individual tags.

This gives a list that maps from tag names to lists of formatting
expressions.  The car of each element gives a tag name and the
cdr gives a list of Elisp expressions that modify the tag.  If
the list is empty, the tag will simply be hidden.  Otherwise,
each expression will be evaluated in order: for the first
expression, the variable `tag' will be bound to the tag name; for
each later expression, the variable `tag' will be bound to the
result of the previous expression.  In this way, each expression
can build on the formatting performed by the previous expression.
The result of the last expression will displayed in place of the
tag.

For example, to replace a tag with another string, simply use
that string as a formatting expression.  To change the foreground
of a tag to red, use the expression
  (propertize tag 'face '(:foreground \"red\"))

See also `notmuch-tag-format-image', which can help replace tags
with images."

  :group 'notmuch-search
  :group 'notmuch-show
  :type '(alist :key-type (string :tag "Tag")
		:extra-offset -3
		:value-type
		(radio :format "%v"
		       (const :tag "Hidden" nil)
		       (set :tag "Modified"
			    (string :tag "Display as")
			    (list :tag "Face" :extra-offset -4
				  (const :format "" :inline t
					 (propertize tag 'face))
				  (list :format "%v"
					(const :format "" quote)
					custom-face-edit))
			    (list :format "%v" :extra-offset -4
				  (const :format "" :inline t
					 (notmuch-tag-format-image-data tag))
				  (choice :tag "Image"
					  (const :tag "Star"
						 (notmuch-tag-star-icon))
					  (const :tag "Empty star"
						 (notmuch-tag-star-empty-icon))
					  (const :tag "Tag"
						 (notmuch-tag-tag-icon))
					  (string :tag "Custom")))
			    (sexp :tag "Custom")))))

(defun notmuch-tag-format-image-data (tag data)
  "Replace TAG with image DATA, if available.

This function returns a propertized string that will display image
DATA in place of TAG.This is designed for use in
`notmuch-tag-formats'.

DATA is the content of an SVG picture (e.g., as returned by
`notmuch-tag-star-icon')."
  (propertize tag 'display
	      `(image :type svg
		      :data ,data
		      :ascent center
		      :mask heuristic)))

(defun notmuch-tag-star-icon ()
  "Return SVG data representing a star icon.
This can be used with `notmuch-tag-format-image-data'."
"<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>
<svg version=\"1.1\" width=\"16\" height=\"16\">
  <g transform=\"translate(-242.81601,-315.59635)\">
    <path
       d=\"m 290.25762,334.31206 -17.64143,-11.77975 -19.70508,7.85447 5.75171,-20.41814 -13.55925,-16.31348 21.19618,-0.83936 11.325,-17.93675 7.34825,19.89939 20.55849,5.22795 -16.65471,13.13786 z\"
       transform=\"matrix(0.2484147,-0.02623394,0.02623394,0.2484147,174.63605,255.37691)\"
       style=\"fill:#ffff00;fill-rule:evenodd;stroke:#000000;stroke-width:1px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1\" />
  </g>
</svg>")

(defun notmuch-tag-star-empty-icon ()
  "Return SVG data representing an empty star icon.
This can be used with `notmuch-tag-format-image-data'."
  "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>
<svg version=\"1.1\" width=\"16\" height=\"16\">
  <g transform=\"translate(-242.81601,-315.59635)\">
    <path
       d=\"m 290.25762,334.31206 -17.64143,-11.77975 -19.70508,7.85447 5.75171,-20.41814 -13.55925,-16.31348 21.19618,-0.83936 11.325,-17.93675 7.34825,19.89939 20.55849,5.22795 -16.65471,13.13786 z\"
       transform=\"matrix(0.2484147,-0.02623394,0.02623394,0.2484147,174.63605,255.37691)\"
       style=\"fill:#d6d6d1;fill-opacity:1;fill-rule:evenodd;stroke:#000000;stroke-width:1px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1\" />
  </g>
</svg>")

(defun notmuch-tag-tag-icon ()
  "Return SVG data representing a tag icon.
This can be used with `notmuch-tag-format-image-data'."
  "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>
<svg version=\"1.1\" width=\"16\" height=\"16\">
  <g transform=\"translate(0,-1036.3622)\">
    <path
       d=\"m 0.44642857,1040.9336 12.50000043,0 2.700893,3.6161 -2.700893,3.616 -12.50000043,0 z\"
       style=\"fill:#ffff00;fill-opacity:1;fill-rule:evenodd;stroke:#000000;stroke-width:0.25;stroke-linecap:butt;stroke-linejoin:miter;stroke-miterlimit:4;stroke-opacity:1\" />
  </g>
</svg>")

(defun notmuch-tag-format-tag (tag)
  "Format TAG by looking into `notmuch-tag-formats'."
  (let ((formats (assoc tag notmuch-tag-formats)))
    (cond
     ((null formats)		;; - Tag not in `notmuch-tag-formats',
      tag)			;;   the format is the tag itself.
     ((null (cdr formats))	;; - Tag was deliberately hidden,
      nil)			;;   no format must be returned
     (t				;; - Tag was found and has formats,
      (let ((tag tag))		;;   we must apply all the formats.
	(dolist (format (cdr formats) tag)
	  (setq tag (eval format))))))))

(defun notmuch-tag-format-tags (tags)
  "Return a string representing formatted TAGS."
  (notmuch-combine-face-text-property-string
   (mapconcat #'identity
	      ;; nil indicated that the tag was deliberately hidden
	      (delq nil (mapcar #'notmuch-tag-format-tag tags))
	      " ")
   'notmuch-tag-face
   t))

(defcustom notmuch-before-tag-hook nil
  "Hooks that are run before tags of a message are modified.

'tags' will contain the tags that are about to be added or removed as
a list of strings of the form \"+TAG\" or \"-TAG\".
'query' will be a string containing the search query that determines
the messages that are about to be tagged"

  :type 'hook
  :options '(notmuch-hl-line-mode)
  :group 'notmuch-hooks)

(defcustom notmuch-after-tag-hook nil
  "Hooks that are run after tags of a message are modified.

'tags' will contain the tags that were added or removed as
a list of strings of the form \"+TAG\" or \"-TAG\".
'query' will be a string containing the search query that determines
the messages that were tagged"
  :type 'hook
  :options '(notmuch-hl-line-mode)
  :group 'notmuch-hooks)

(defvar notmuch-select-tag-history nil
  "Variable to store minibuffer history for
`notmuch-select-tag-with-completion' function.")

(defvar notmuch-read-tag-changes-history nil
  "Variable to store minibuffer history for
`notmuch-read-tag-changes' function.")

(defun notmuch-tag-completions (&rest search-terms)
  "Return a list of tags for messages matching SEARCH-TERMS.

Returns all tags if no search terms are given."
  (if (null search-terms)
      (setq search-terms (list "*")))
  (split-string
   (with-output-to-string
     (with-current-buffer standard-output
       (apply 'call-process notmuch-command nil t
	      nil "search" "--output=tags" "--exclude=false" search-terms)))
   "\n+" t))

(defun notmuch-select-tag-with-completion (prompt &rest search-terms)
  (let ((tag-list (apply #'notmuch-tag-completions search-terms)))
    (completing-read prompt tag-list nil nil nil 'notmuch-select-tag-history)))

(defun notmuch-read-tag-changes (current-tags &optional prompt initial-input)
  "Prompt for tag changes in the minibuffer.

CURRENT-TAGS is a list of tags that are present on the message or
messages to be changed.  These are offered as tag removal
completions.  CURRENT-TAGS may contain duplicates.  PROMPT, if
non-nil, is the query string to present in the minibuffer.  It
defaults to \"Tags\".  INITIAL-INPUT, if non-nil, will be the
initial input in the minibuffer."

  (let* ((all-tag-list (notmuch-tag-completions))
	 (add-tag-list (mapcar (apply-partially 'concat "+") all-tag-list))
	 (remove-tag-list (mapcar (apply-partially 'concat "-") current-tags))
	 (tag-list (append add-tag-list remove-tag-list))
	 (prompt (concat (or prompt "Tags") " (+add -drop): "))
	 (crm-separator " ")
	 ;; By default, space is bound to "complete word" function.
	 ;; Re-bind it to insert a space instead.  Note that <tab>
	 ;; still does the completion.
	 (crm-local-completion-map
	  (let ((map (make-sparse-keymap)))
	    (set-keymap-parent map crm-local-completion-map)
	    (define-key map " " 'self-insert-command)
	    map)))
    (delete "" (completing-read-multiple
		prompt
		;; Append the separator to each completion so when the
		;; user completes a tag they can immediately begin
		;; entering another.  `completing-read-multiple'
		;; ultimately splits the input on crm-separator, so we
		;; don't need to strip this back off (we just need to
		;; delete "empty" entries caused by trailing spaces).
		(mapcar (lambda (tag-op) (concat tag-op crm-separator)) tag-list)
		nil nil initial-input
		'notmuch-read-tag-changes-history))))

(defun notmuch-update-tags (tags tag-changes)
  "Return a copy of TAGS with additions and removals from TAG-CHANGES.

TAG-CHANGES must be a list of tags names, each prefixed with
either a \"+\" to indicate the tag should be added to TAGS if not
present or a \"-\" to indicate that the tag should be removed
from TAGS if present."
  (let ((result-tags (copy-sequence tags)))
    (dolist (tag-change tag-changes)
      (let ((op (string-to-char tag-change))
	    (tag (unless (string= tag-change "") (substring tag-change 1))))
	(case op
	  (?+ (unless (member tag result-tags)
		(push tag result-tags)))
	  (?- (setq result-tags (delete tag result-tags)))
	  (otherwise
	   (error "Changed tag must be of the form `+this_tag' or `-that_tag'")))))
    (sort result-tags 'string<)))

(defconst notmuch-tag-argument-limit 1000
  "Use batch tagging if the tagging query is longer than this.

This limits the length of arguments passed to the notmuch CLI to
avoid system argument length limits and performance problems.")

(defun notmuch-tag (query tag-changes)
  "Add/remove tags in TAG-CHANGES to messages matching QUERY.

QUERY should be a string containing the search-terms.
TAG-CHANGES is a list of strings of the form \"+tag\" or
\"-tag\" to add or remove tags, respectively.

Note: Other code should always use this function alter tags of
messages instead of running (notmuch-call-notmuch-process \"tag\" ..)
directly, so that hooks specified in notmuch-before-tag-hook and
notmuch-after-tag-hook will be run."
  ;; Perform some validation
  (mapc (lambda (tag-change)
	  (unless (string-match-p "^[-+]\\S-+$" tag-change)
	    (error "Tag must be of the form `+this_tag' or `-that_tag'")))
	tag-changes)
  (unless (null tag-changes)
    (run-hooks 'notmuch-before-tag-hook)
    (if (<= (length query) notmuch-tag-argument-limit)
	(apply 'notmuch-call-notmuch-process "tag"
	       (append tag-changes (list "--" query)))
      ;; Use batch tag mode to avoid argument length limitations
      (let ((batch-op (concat (mapconcat #'notmuch-hex-encode tag-changes " ")
			      " -- " query)))
	(notmuch-call-notmuch-process :stdin-string batch-op "tag" "--batch")))
    (run-hooks 'notmuch-after-tag-hook)))

(defun notmuch-tag-change-list (tags &optional reverse)
  "Convert TAGS into a list of tag changes.

Add a \"+\" prefix to any tag in TAGS list that doesn't already
begin with a \"+\" or a \"-\". If REVERSE is non-nil, replace all
\"+\" prefixes with \"-\" and vice versa in the result."
  (mapcar (lambda (str)
	    (let ((s (if (string-match "^[+-]" str) str (concat "+" str))))
	      (if reverse
		  (concat (if (= (string-to-char s) ?-) "+" "-")
			  (substring s 1))
		s)))
	  tags))


;;

(provide 'notmuch-tag)

;; Local Variables:
;; byte-compile-warnings: (not cl-functions)
;; End:
